File: lib/src/main/java/com/github/teren4m/base/mvvm/ListLiveData.java
package com.github.teren4m.base.mvvm;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;

import java.util.ArrayList;
import java.util.List;

public class ListLiveData<T> extends LiveData<List<T>> {

    public ListLiveData() {
        this(100);
    }

    public ListLiveData(int initialCapacity) {
        setValue(new ArrayList<T>(initialCapacity));
    }

    public void add(List<T> items) {
        if (getValue() != null && items != null) {
            getValue().addAll(items);
            setValue(getValue());
        }
    }

    public void add(T item) {
        if (getValue() != null && item != null) {
            getValue().add(item);
            setValue(getValue());
        }
    }

    public void clear() {
        if (getValue() != null) {
            getValue().clear();
            setValue(getValue());
        }
    }

    @Override
    public void setValue(List<T> value) {
        super.setValue(value);
    }

    @Nullable
    @Override
    public List<T> getValue() {
        return super.getValue();
    }
}

File: lib/src/main/java/com/github/teren4m/base/mvvm/ListLiveData.kt
package com.github.teren4m.base.mvvm

import androidx.lifecycle.LiveData
import java.util.*

class ListLiveData<T : Any> @JvmOverloads constructor(initialCapacity: Int = 100) : LiveData<List<T>>() {

    init {
        value = ArrayList(initialCapacity)
    }

    fun add(items: List<T>) {
        getList().apply {
            addAll(items)
            postValue(value)
        }
    }

    fun add(item: T) {
        getList().apply {
            add(item)
            postValue(value)
        }
    }

    fun clear() {
        getList().apply {
            clear()
            postValue(value)
        }
    }

    private fun getList(): ArrayList<T> =
        value as ArrayList<T>

}